### PR TITLE
Order and stepsize control RichardsonEuler

### PR DESCRIPTION
**Pull request by saurabhkgp21**
This PR is still in its initial stage. There are still bugs in this implementation.
@ChrisRackauckas is this a correct way to go on for Variable Order?

### CODE REVIEW

File: src/caches/extrapolation_caches.jl [hunk 1]
```diff
@@ -1,31 +1,53 @@
...
   fsalfirst = zero(rate_prototype)
-  tab = RichardsonEulerConstantCache()
-  T = fill(zeros(eltype(u), size(u)), (tab.m, tab.m))
-  RichardsonEulerCache(u,uprev,tmp,k,fsalfirst,tab,T)
+  order = 1
+  order_max = 9
+  dtproposeold = zero(dt)
+  dtpropose = zero(dt)
+  T = fill(zeros(eltype(u), size(u)), (order_max, order_max))
+  last_work = zero(dt)
```
Comment (ChrisRackauckas) [inline src/caches/extrapolation_caches.jl]: prev. notice that the rest of the package uses `prev` for previous.

File: src/integrators/integrator_utils.jl [hunk 2]
```diff
@@ -198,6 +199,23 @@ function step_reject_controller!(integrator,alg::JVODE)
   integrator.dt *= integrator.qold
 end
 
+function stepsize_controller!(integrator,alg::RichardsonEuler)
```
Comment (ChrisRackauckas) [inline src/integrators/integrator_utils.jl]: this is just a standard proportional controller? If so, no reason to just copy paste the code. Just change how it dispatches.

File: src/perform_step/extrapolation_perform_step.jl [hunk 3]
```diff
@@ -48,17 +48,22 @@ function initialize!(integrator,cache::RichardsonEulerConstantCache)
   integrator.fsallast = zero(integrator.fsalfirst)
   integrator.k[1] = integrator.fsalfirst
   integrator.k[2] = integrator.fsallast
+
+  cache.order = 3
```
Comment (ChrisRackauckas) [inline src/perform_step/extrapolation_perform_step.jl]: ?

File: src/perform_step/extrapolation_perform_step.jl [hunk 4]
```diff
@@ -48,17 +48,22 @@ function initialize!(integrator,cache::RichardsonEulerConstantCache)
...
-  T = zeros(typeof(uprev), (m,m))
+  A = 1 + 1
+
   T[1,1] = u
+  dtpropose = dt #doubt but error_1 isn't defined
+  work = A/dtpropose
 
   halfdt = dt/2
-  for i in 2:m
+  for i in 2:order_max
```
Comment (ChrisRackauckas) [inline src/perform_step/extrapolation_perform_step.jl]: order?

File: src/perform_step/extrapolation_perform_step.jl [hunk 5]
```diff
@@ -73,14 +78,96 @@ function perform_step!(integrator,cache::RichardsonEulerConstantCache,repeat_ste
...
+              integrator.k[2] = integrator.fsallast
+              return
+          else
+              if integrator.EEst > 4*(dt/halfdt)^4
+                  cache.order = order - 1
+                  cache.dtpropose = dtpropose
+                  return
+              end
+          end
+      elseif i == order
```
Comment (ChrisRackauckas) [inline src/perform_step/extrapolation_perform_step.jl]: it looks like this only calculates if adaptive and i == order? That doesn't seem correct.

File: src/caches/extrapolation_caches.jl [hunk 6]
```diff
@@ -1,31 +1,53 @@
...
 
-struct RichardsonEulerConstantCache <: OrdinaryDiffEqConstantCache
-  m::Int #Generalise type of m
-  # Can add different order of n_{i} here
-  function RichardsonEulerConstantCache()
-    m = 4
-    new(m)
-  end
+@cache mutable struct RichardsonEulerConstantCache{dtType,arrayType} <: OrdinaryDiffEqConstantCache
+  dtproposeold::dtType
```
[the remark this answers is not in the document]
Comment (saurabhkgp21) [reply src/caches/extrapolation_caches.jl]: @ChrisRackauckas I had used dt argument of alg_cache() to initialise `prevdtpropose`. Shouldn't it match with `tType` by itself?

File: src/perform_step/extrapolation_perform_step.jl (hunk 4 above)
Comment (saurabhkgp21) [reply src/perform_step/extrapolation_perform_step.jl]: size of `T` is `(max_order, max_order)`, hence I had written something like this in my pseudo code. Now I had changed this with `size(T)[0]`. Is this fine?

File: src/perform_step/extrapolation_perform_step.jl (hunk 5 above)
Comment (saurabhkgp21) [reply src/perform_step/extrapolation_perform_step.jl]: @ChrisRackauckas  If step hasn't converged for `line == k - 1` it goes on to check for convergence in `line == k`. Here I have used order as `k`. Though now I have renamed order as cur_order for more clarity.
Am I missing something here?

File: src/integrators/integrator_utils.jl (hunk 2 above)
Comment (saurabhkgp21) [reply src/integrators/integrator_utils.jl]: @ChrisRackauckas Here controller modifies `dt` according to calculated `dtpropose` in `perform_step`.

Comment (ChrisRackauckas) [reply src/integrators/integrator_utils.jl]: ahh it changed.

File: src/perform_step/extrapolation_perform_step.jl (hunk 5 above)
Comment (ChrisRackauckas) [reply src/perform_step/extrapolation_perform_step.jl]: Don't use `k` for order, it already has a meaning. Just name it `order`

File: src/algorithms.jl [hunk 7]
```diff
@@ -31,7 +32,13 @@ ExplicitRK(;tableau=ODE_DEFAULT_TABLEAU) = ExplicitRK(tableau)
 @inline trivial_limiter!(u, f, t) = nothing
 
 struct Euler <: OrdinaryDiffEqAlgorithm end
-struct RichardsonEuler <: OrdinaryDiffEqAlgorithm end
+struct RichardsonEuler <: OrdinaryDiffEqExtrapolationVarOrderVarStepAlgorithm
+  max_order::Int
+  min_order::Int
+  init_order::Int
+end
+RichardsonEuler(;max_order=9,min_order=1,init_order=9) = RichardsonEuler(max_order,min_order,init_order)
```
Comment (ChrisRackauckas) [inline src/algorithms.jl]: init at order 9 might be high. I'd put it to 5

Comment (saurabhkgp21) [reply src/algorithms.jl]: In the initialising phase, work per step is checked for k = 1 to max_order. And the order was going as high as 7 in some cases. That's why I had put it as 9. If there can be other problems due to high init_order, I'll decrease this one.

Comment (ChrisRackauckas) [reply src/algorithms.jl]: The problem you were testing on was completely perfect for high order. Not all problems are like that.

Comment (saurabhkgp21) [reply src/algorithms.jl]: okay.. will set init_order as 5 and leave max_order as 9. Will it be fine?

Comment (ChrisRackauckas) [reply src/algorithms.jl]: yes

Comment (saurabhkgp21) [reply src/algorithms.jl]: done

**Comment by ChrisRackauckas** [comment]
Yeah, I took a good look through it. It looks good. We will want a version that uses GBS to parallelize on though, since this method won't be quite as efficient. But as a method it looks complete. Is there a reason why this is still WIP?

**Comment by saurabhkgp21** [comment]
Changed the title now. Yeah GBS will give better results upon parallelization than this one. We can also parallelize this one first since order and stepsize control is completed. After that, I can implement parallelize version of GBS.

**Pull request merged**